File: 1498-find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def getTargetCopy(self, original: TreeNode, cloned: TreeNode, target: TreeNode) -> TreeNode:
        self._target = target
        self.res = None

        def traverse(original, cloned):
            if original is None or self.res is not None:
                return
            
            if original == target:
                self.res = cloned
                return
            
            traverse(original.left,  cloned.left)
            traverse(original.right, cloned.right)
        
        traverse(original, cloned)
        return self.res

File: 1498-find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public final TreeNode getTargetCopy(final TreeNode original, final TreeNode cloned, final TreeNode target) {
        if (original == null) {
            return null;
        }
        // 找到目标节点
        if (target == original) {
            return cloned;
        }
        // 去左子树找
        TreeNode left = getTargetCopy(original.left, cloned.left, target);
        if (left != null) {
            return left;
        }
        // 左子树找不到的话去右子树找
        return getTargetCopy(original.right, cloned.right, target);
    }
}